external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp

```diff
@@ -513,54 +554,159 @@ class TosaPartitionPass : public TosaPartitionBase<TosaPartitionPass> {
+              // If all the users of any resultNode are in trailingOps, there's
               // no need to return it so remove from resultNodes.  Finally,
               // add all userOp's results to resultNodes.
               if (!skip) {
-                secondOps.insert(userOp);
+                if (isa<tosa::TransposeOp>(userOp)) {
+                  specialCaseForTranspose(userOp, trailingOps);
+                }
+                // General case.
+                trailingOps.insert(userOp);
                 worklist.push_back(userOp);
-                for (Value opnd : userOp->getOperands()) {
-                  if (!resultNodes.contains(opnd)) {
+                for (Value opnd : userOp->getOperands())
+                  if (!resultNodes.contains(opnd))
                     inputNodes.insert(opnd);
-                  }
-                }
-                for (const Value &val : resultNodes) {
+                for (const Value &val : resultNodes)
                   if (llvm::all_of(val.getUsers(), [&](Operation *u) {
-                        return secondOps.contains(u);
-                      })) {
+                        return trailingOps.contains(u);
+                      }))
                     resultNodes.remove(val);
-                  }
-                }
-                for (auto res : userOp->getResults()) {
+                for (auto res : userOp->getResults())
                   resultNodes.insert(res);
-                }
               }
             }
           }
         }
 
         // Make the outlined function from the ops we've gathered.
-        outlineConvPartOps(convOp, secondOps.getArrayRef(), frontOps,
-                           inputNodes.getArrayRef(), resultNodes.getArrayRef(),
+        outlineConvPartOps(anchorOp, trailingOps.getArrayRef(),
+                           leadingOps.getArrayRef(), inputNodes.getArrayRef(),
+                           resultNodes.getArrayRef(),
                            std::string(func.getSymName()) + strCount,
-                           candidates);
+                           attributeName(), candidates);
         // Outlining will erase nodes and thus perturb the walk, so
         // signal interrupted to exit it and restart.
         return WalkResult::interrupt();
       };
 
-      // Walk until we've outlined all the Conv2D ops we can.
+      // Walk until we've outlined all the anchor ops we can.
       while (func.walk(callback).wasInterrupted()) {
       }
     }
   }
 };
 
+class TosaPartitionPassWithOptions
+  : public TosaPartitionPass {
+public:
+  TosaPartitionPassWithOptions() = default;
+  TosaPartitionPassWithOptions(ArrayRef<std::string> anchorOps_,
+                               std::string attrName, bool trailingOnly_) {
+    anchorOps = anchorOps_;
+    attributeNameOpt = attrName;
+    trailingOnly = trailingOnly_;
+  }
+
+  bool isAnchorOp(Operation *op) override {
+    if (anchorOps.empty())   // ListOption doesn't have a default value.
```

Comment:
Have to test it here, because it'll always be empty at pass-creation time.  I think that's an artifact of registering the passes before the command line is parsed.  In the last version, I did this in runOnOperation();  I'd have to override that method to do the same with this version.

Either way, the anchor-op test is always a list-membership test in this version.